algebird-core/src/main/scala/com/twitter/algebird/Preparer.scala

```diff
@@ -0,0 +1,134 @@
+package com.twitter.algebird
+
+trait Preparer[A, T, +This[A, T] <: Preparer[A, T, This]] extends HasMonoidAggregate[A, T] {
+  def map[U](fn: T => U): This[A, U]
+
+  def flatMap[U](fn: T => TraversableOnce[U]): FlatMapPreparer[A, U]
+
+  def flatten[U](implicit ev: <:<[T, TraversableOnce[U]]) = flatMap(ev)
+
+  def filter(fn: T => Boolean) = flatMap{ t => if (fn(t)) Some(t) else None }
+
+  //really ought to generate N versions of this...
+  def split[B1, B2, C1, C2](fn: This[A, T] => (Aggregator[A, B1, C1], Aggregator[A, B2, C2])): Aggregator[A, (B1, B2), (C1, C2)] = {
+    val (a1, a2) = fn(this.asInstanceOf[This[A, T]])
+    a1.join(a2)
+  }
+}
+
+object Preparer {
+  def apply[A] = MapPreparer[A, A](identity)
```

Comment:
I wonder if it is worth making `IdentityPreparer[A]` that does not bother calling compose. It might help perf a bit since the jit does not have to realize identity does nothing.